metagraph/src/graph/representation/hash/dbg_sshash.cpp

```diff
@@ -0,0 +1,494 @@
+    call_outgoing_kmers(node, [&](auto child, char) { callback(child); });
+}
+
+void DBGSSHash::adjacent_incoming_nodes(node_index node,
+                                        const std::function<void(node_index)>& callback) const {
+    assert(node > 0 && node <= num_nodes());
+    call_incoming_kmers(node, [&](auto parent, char) { callback(parent); });
+}
+
+void DBGSSHash::call_outgoing_kmers(node_index node,
+                                    const OutgoingEdgeCallback& callback) const {
+    assert(node > 0 && node <= num_nodes());
+    std::string kmer = get_node_sequence(node);
+    auto call_impl = [&](const auto &dict) {
+        using kmer_t = typename std::remove_reference<decltype(dict)>::type::kmer_t;
+        sshash::neighbourhood<kmer_t> nb = dict.kmer_forward_neighbours(kmer.c_str(), mode_ == CANONICAL);
+        for (size_t i = 0; i < kmer_t::alphabet_size; i++) {
+            if (nb.forward[i].kmer_id != sshash::constants::invalid_uint64) {
+                node_index next = sshash_to_graph_index(nb.forward[i].kmer_id);
+                callback(nb.forward[i].kmer_orientation ? reverse_complement(next) : next, kmer_t::alphabet[i]);
+            }
+        }
+    };
+
+    if (auto dict64 = dynamic_cast<Dictionary<Kmer64>*>(dict_.get())) {
+        call_impl(*dict64);
+    } else if (auto dict128 = dynamic_cast<Dictionary<Kmer128>*>(dict_.get())) {
+        call_impl(*dict128);
+    } else if (auto dict256 = dynamic_cast<Dictionary<Kmer256>*>(dict_.get())) {
+        call_impl(*dict256);
+    } else {
+        common::logger->error("");
+        throw std::runtime_error("k fail");
+    }
+}
+
+void DBGSSHash::call_incoming_kmers(node_index node,
+                                    const IncomingEdgeCallback& callback) const {
+    assert(node > 0 && node <= num_nodes());
+    std::string kmer = get_node_sequence(node);
+    auto call_impl = [&](const auto &dict) {
+        using kmer_t = typename std::remove_reference<decltype(dict)>::type::kmer_t;
+        sshash::neighbourhood<kmer_t> nb = dict.kmer_backward_neighbours(kmer.c_str(), mode_ == CANONICAL);
+        for (size_t i = 0; i < kmer_t::alphabet_size; i++) {
+            if (nb.backward[i].kmer_id != sshash::constants::invalid_uint64) {
+                node_index prev = sshash_to_graph_index(nb.backward[i].kmer_id);
+                callback(nb.backward[i].kmer_orientation ? reverse_complement(prev) : prev, kmer_t::alphabet[i]);
+            }
+        }
+    };
+
+    if (auto dict64 = dynamic_cast<Dictionary<Kmer64>*>(dict_.get())) {
+        call_impl(*dict64);
+    } else if (auto dict128 = dynamic_cast<Dictionary<Kmer128>*>(dict_.get())) {
+        call_impl(*dict128);
+    } else if (auto dict256 = dynamic_cast<Dictionary<Kmer256>*>(dict_.get())) {
+        call_impl(*dict256);
+    } else {
+        common::logger->error("");
+        throw std::runtime_error("k fail");
+    }
+}
+
+void DBGSSHash::call_outgoing_kmers_with_rc(
```

Comment:
I'd also prefer to govern this (and the next function) via `template<bool rc = false>` too.